### DISCUSSION TITLE
Docker desktop VM IP

### DISCUSSION THREAD

**Discussion by johnpangalos** [moby/moby#45489] [Q&A]
I haven't found any documentation on this and I was wondering how to tell what the IP of the docker vm created by Docker Desktop is going to be.

On my Arm Mac running Docker Engine v23.0.5 I get that it's `172.17.42.1` but my colleague with an Intel Mac running Docker Engine v20.10.14 get `192.168.65.2`.

We have some home made scripts that depend on this functionality and I was wondering if there was someway we could tell which IP it's going to be before running the script.

Any help with this would be greatly appreciated.

**Comment by blueprismo** [comment]
Hello @johnpangalos ,
I found a "similar" scenario and stumbled upon your question, I agree there are hidden parts of the documentation about which is the default CIDR that docker uses, some google dorks suggest that it uses the `172.17.0.0/16`, but as your friend got another one... how can this happen?
There are multiple configuration set ups and I think you may have missed some context... Are those two scenarios a fresh-new installation? Telling so because your friend IP address looks like it's on the https://docs.docker.com/network/drivers/macvlan/  network driver.
What may be helpful is to use the `docker network inspect bridge` if you all want to use the "default" bridge network.
```
# Out put of "docker network inspect bridge"
[
    {
        "Name": "bridge",
        "IPAM": {
            "Config": [
                {
                    "Subnet": "172.17.0.0/16",
                    "Gateway": "172.17.0.1"
                }
            ]
[...]
```
If it's the bridge network that has the undesired CIDR range, you can manually specify it in your `/etc/docker/daemon.json` file with the  [bip range](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon) values you wish.

Hope I could help!

**Reply by polarathene** [reply]
Just to clarify, there is a difference between:
- Windows / macOS VM guest (_that docker runs internally in_) network IP
- Docker daemon default address pools for creating networks for containers to have IP assigned and bridge to the Docker host network interface (_VM guest network in this case_).

Since @blueprismo seems to be referring to the latter, I've answered that below where those addresses come from. 

VM guests (and Docker daemon) will often use IP subnets from the [private address range](https://en.wikipedia.org/wiki/Private_network), such as the one Docker Desktop is managing to run Docker inside of. Docker will implicitly choose a subnet from the pools below that don't overlap with one already in use (such as by the VM guest) AFAIK.

---

> I agree there are hidden parts of the documentation about which is the default CIDR that docker uses, some google dorks suggest that it uses the `172.17.0.0/16`, but as your friend got another one... how can this happen?

There is a default pool, at the time this discussion was made I think the information was missing but it did get documented in the [IPv6 page update earlier this year](https://github.com/docker/docs/pull/17176): https://docs.docker.com/config/daemon/ipv6/#dynamic-ipv6-subnet-allocation

```json
{
  "default-address-pools": [
    { "base": "172.17.0.0/16", "size": 16 },
    { "base": "172.18.0.0/16", "size": 16 },
    { "base": "172.19.0.0/16", "size": 16 },
    { "base": "172.20.0.0/14", "size": 16 },
    { "base": "172.24.0.0/14", "size": 16 },
    { "base": "172.28.0.0/14", "size": 16 },
    { "base": "192.168.0.0/16", "size": 20 }
  ]
}
```

It's not exactly the best place for it to be discovered, ~~and I'm not 100% sure if that is the full subnet pool as I recall another private IPv4 address range `10.0.0.0/8` was possibly also covered? Maybe it no longer is, one would need to look up the source code 😅 ~~

Here was a previous issue about documenting this properly: https://github.com/docker/docs/issues/8663
- Comments link/reference the location of the source at the time in the [`moby/libnetwork`](https://github.com/moby/libnetwork/blob/3797618f9a38372e8107d8c06f6ae199e1133ae8/ipamutils/utils.go#L18-L21) repo.
- However since then `libnetwork` was relocated into `moby/moby` repo, so [this is the new location](https://github.com/moby/moby/blob/ce1ee98aba41e117c59bf17a108191eed14e9b9c/libnetwork/ipamutils/utils.go#L18-L27):

https://github.com/moby/moby/blob/ce1ee98aba41e117c59bf17a108191eed14e9b9c/libnetwork/ipamutils/utils.go#L18-L27

So that's the CIDR of `10.0.0.0/8` split into many smaller subnets of `10.x.y.z/24`, same logic for the other pools `base` + `size` config ([like described here](https://github.com/docker/docs/issues/8663#issuecomment-956438889)).

---

The documentation was [meant to be improved on this](https://github.com/docker/docs/pull/17176#discussion_r1217990449) and I had [raised it as a concern during the IPv6 rewrite review](https://github.com/docker/docs/pull/17176#discussion_r1211135680).

I've [opened a new issue on the docs repo](https://github.com/docker/docs/issues/18782) to get this better documented 👍
